### PR TITLE
A Jupyter notebook and some html+js to visually explore graphs. 

### PR DESCRIPTION
**Pull request by gerasdf**
We'll have to plug it with Cuis, but it's already useful

to use run jupyter-notebook in the project's root path, and from the browser open Graph.ipynb, then evaluate a few lines. You need to have python, tensorflow (for python) and numpy installed (as well as jupyter-notebooks)

### CODE REVIEW

**Comment by gerasdf** [comment]
btw, this does not work in Firefox, because Firefox does not support html link includes (https://developer.mozilla.org/en-US/docs/Web/Web_Components/HTML_Imports) and this implementation uses it... not sure if it's strictly necessary, but it does use it

**Comment by gerasdf** [comment]
Here a bunch of screenshots on how I'm using the graph visualization tool
in this branch. First (0), only the forward graph, pretty simple. I think
it's the same we are building in Cuis.

1 to 4 are different shots of the complete graph with learning steps.

1 - All collapsed
2 - SoftMaxCrossEntropyWithLogits gradient computation
3 - Bias addition gradient computation
4 - Weights matrix multiplication matrix computation
5 - The original example, not using *Sparse* SoftMax, see the complexity

I'm using this to understand how to manually add the gradient operations to
the example graphs. Compare to 5, which doesn't use the Sparse version of
SoftMax, see all the extra complexity for reshaping tensors and processing
the training batches.


[image: Inline image 1][image: Inline image 2][image: Inline image 3][image:
Inline image 4][image: Inline image 5]

**Comment by jburroni** [comment]
the images in the previous comment are missing

**Comment by gerasdf** [comment]
well... of course, I confused all attachments... here again:

SoftMax-forward.png - Only Forward Graph

SoftMax-Learning-0-Collapsed.png - Learning Graph, Collapsed
SoftMax-Learning-1-Gradients.png - - Learning Graph, Gradients computation
subgraph open
SoftMax-Learning-2-MeanGraph.png - Learning Graph, Mean gradient subgraph
computation
SoftMax-Learning-3-Add.png - Learning Graph, Bias addition gradient
computation
SoftMax-Learning-4-MatMul.png - Learning Graph,  Weights matrix
multiplication gradient computation
SoftMax-Learning-5-SoftMax.png - Learning Graph, SoftMax gradient
computation (the operation already computes its gradient)

NoSparseSoftMax.png - Learning Graph, not using the Sparse SoftMax, a lot
of complexity slicing the batches

[image: Inline image 1][image: Inline image 2][image: Inline image 3][image:
Inline image 4][image: Inline image 5][image: Inline image 7][image: Inline
image 8][image: Inline image 6]

**Comment by jburroni** [comment]
As a side note. The port is hardcoded. If the images do not show in the notebook, one should check if the notebook's port is the same as in the HTML
